### ISSUE TITLE
Plan for ASG/LF-MMI

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/k2#558]
My proposal is to have the phone LM have normalized scores and be trained only from the numerator alignments.
```
 phone_list = [1,2,3,...];
 # note: phone_list should be just the real phones, not including disambig symbols.
 # We would need to find a way to distinguish the disambig symbols from the real phones.
 # P has a start-state and a state for each phone 0, 1, ....; and each of the above-mentioned states
 # has a transition to the state for each phone and also to the final-state
 P = create_bigram_phone_lm(phone_list)
 
 # The next line assumes model.P_scores was initialized to the size (num_phones+1)*(num_phones+1).
 # We could almost-equivalently do model.P_scores = P.scores, but that wouldn't work correctly
 # when loading the model from disk, it would reset P.scores.
 P.scores = model.P_scores 
```
Then on each minibatch, we'd do:
```
  den_graph = k2.intersect(P, ctc_topo).detach()  # want only deriv from numerator
```
 ... and we'd also need to intersect the numerator graphs with P before combining them
with ctc_topo; and do this in a differentiable way; this of course would need to be done
without caching so the autograd would work correcty.  This won't directly affect the 
training much (although it will produce more interpretable scores); the point is mostly the
derivative information which can train P from the training transcripts.  The motivation
for training P as if it were an LM is so that our model will "expect" the language model
to be like a real LM, so it will work well with the G that we estimate from data.

**Comment by csukuangfj** [comment]
I will try to implement it tomorrow.

**Comment by csukuangfj** [comment]
```
  den_graph = k2.intersect(P, ctc_topo).detach()  # want only deriv from numerator
```

Should we set `treat_epsilons_specially` to `False`?

**Comment by csukuangfj** [comment]
I think `0` means blank here, not epsilon and we are doing `P intersect ctc_topo`, **NOT** `ctc_topo intersect P`.

**Comment by danpovey** [comment]
No, treat_epsilons_specially should be true when intersecting P with ctc_topo, because the epsilons on the side of ctc_topo that we're doing the intersection with are actually real epsilons.  It's the ones on the other side that represent blank.

**Comment by csukuangfj** [comment]
Thanks, I see.
